### PR TITLE
[Machine Learning] How to Implement K fold Cross-Validation in Scikit-Learn 

### LINKED ISSUE #5223: [Machine Learning] How to Implement K fold Cross-Validation in Scikit-Learn
### Topic Suggestion
[Machine Learning] How to Implement K fold Cross-Validation in Scikit-Learn

#### Proposed title of the article
How to Implement K fold Cross-Validation in Scikit-Learn

#### Proposed article introduction
In machine learning, the most common approach is to split a dataset into the training set and the validation set. The splitting ratio could be 70:30 or 80:20. The problem with this approach is we are not sure if a good validation accuracy means the model is really good. What if the portion of the data used as validation luckily gave us a good result? If we had used another portion of the dataset as a validation set, would our model still give us a good accuracy? These are some of the things we are not sure of. K fold cross-validation provides a better way of validating our model performance. It evaluates the model using different portions of the dataset as the validation set in various iterations.

The K fold cross-validation technique is helpful when you are training your model on a small dataset. If your dataset is large (maybe over 100,000 samples), you may not have to use K fold cross-validation. The reason is you will have enough data records in your validation set to evaluate the machine learning model. Also, using the K fold cross-validation on a large dataset will take a lot of time.

#### Key takeaways 
What you will learn:
1.	How to use K fold cross-validation to measure the performance of the model.
2.	How to use the decision tree algorithm in Sci-kit Learn.
3.	How to visualize your training accuracy, training recall, training precision and training f1-score on each fold of the dataset.
4.	How to visualize your validation accuracy, validation recall, validation precision and validation f1-score on each fold of the dataset.
5. How to reduce overfitting in a decision trees model.

#### Article quality
This article is in-depth. It will teach people how to visualize a machine learning model performance. They will be able to tell if the model is overfitting or underfitting. I realized that most of the models trained using decision trees overfit. If you can visualize the training and validation performance of the model, you’ll be able to prevent this by adjusting the hyperparameters of your decision tree model.

### PR DESCRIPTION
**Pull request by Inyrkz**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

#### Topic suggestion this closes

This closes #5223

### CODE REVIEW

**Comment by wilkistermumbi** [comment]
@hectorkambow I'd like to review this PR.

**Comment by WanjaMIKE** [comment]
Plagiarism check done

File: content/articles/how-to-implement-k-fold-cross-validation/index.md [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+     'M' 'B' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'M' 'B' 'M' 'B' 'B' 'B' 'B' 'B' 'M'
+     'M' 'B' 'M' 'M' 'B' 'B' 'B' 'B' 'M' 'B' 'M' 'M' 'B' 'B' 'B' 'B' 'M' 'B'
+     'M' 'M' 'B' 'M' 'B' 'M' 'M' 'B' 'B' 'B' 'M' 'M' 'B' 'M' 'M' 'M' 'B' 'B'
+     'B' 'M' 'B' 'B' 'M' 'M' 'B' 'B' 'B' 'M' 'M' 'B' 'B' 'B' 'B' 'M' 'B' 'B'
+     'M' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'M' 'M' 'M' 'B' 'M' 'M' 'B' 'B' 'B'
+     'M' 'M' 'B' 'M' 'B' 'M' 'M' 'B' 'M' 'M' 'B' 'B' 'M' 'B' 'B' 'M' 'B' 'B'
+     'B' 'B' 'M' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'M' 'B' 'B' 'B' 'B' 'M'
+     'M' 'B' 'M' 'B' 'B' 'M' 'M' 'B' 'B' 'M' 'M' 'B' 'B' 'B' 'B' 'M' 'B' 'B'
+     'M' 'M' 'M' 'B' 'M' 'B' 'M' 'B' 'B' 'B' 'M' 'B' 'B' 'M' 'M' 'B' 'M' 'M'
+     'M' 'M' 'B' 'M' 'M' 'M' 'B' 'M' 'B' 'M' 'B' 'B' 'M' 'B' 'M' 'M' 'M' 'M'
```
Comment (wilkistermumbi) [inline content/articles/how-to-implement-k-fold-cross-validation/index.md:82]: All your outputs need to be inside a `bash` command. Go through the whole tutorial and edit accordingly

File: content/articles/how-to-implement-k-fold-cross-validation/index.md [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+```python
+# Plot F1-Score Result
+plot_result(model_name,
+            "F1",
+            "F1 Scores in 5 Folds",
+            decision_tree_result["Training F1 scores"],
+            decision_tree_result["Validation F1 scores"])
+```
+
+![F1 Score Plot of Decision Tree Model](/engineering-education/how-to-implement-k-fold-cross-validation/f1_score_1.png)
```
Comment (wilkistermumbi) [inline content/articles/how-to-implement-k-fold-cross-validation/index.md:326]: Please separate your file names using hyphens and not underscores. It is good for SEO purposes. In addition, some of your file names have `.PNG` in caps. They all need to be in small letters. Change all the file names with the requested changes accordingly.

File: content/articles/how-to-implement-k-fold-cross-validation/index.md [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+- [References](#references)
+
+### Introduction
+K fold cross-validation is a superior technique to validate the performance of our model. It evaluates the model using different chunks of the data set as the validation set. 
+
+We divide our data set into K folds. K represents the number of folds into which you want to split your data. If we use 5-folds, the data set divides into five sections. In different iterations, one part becomes the validation set.
+
+![5 Fold Cross-Validation](/engineering-education/how-to-implement-k-fold-cross-validation/5_fold_cv.jpeg)
+
+[Image Source: Great Learning Blog](https://www.mygreatlearning.com/blog/cross-validation/)
```
Comment (wilkistermumbi) [inline content/articles/how-to-implement-k-fold-cross-validation/index.md:25]: The formatting is slightly wrong. Please refer to this [article](https://github.com/section-engineering-education/engineering-education/pull/5163/files) on line 53 and edit accordingly.

**Comment by wilkistermumbi** [comment]
Hi @Inyrkz Please work on the revisions before we can proceed for final review. Also, please perform a Grammarly check on your tutorial. Let me know when done.

File: content/articles/how-to-implement-k-fold-cross-validation/index.md (hunk 1 above)
Comment (Inyrkz) [reply content/articles/how-to-implement-k-fold-cross-validation/index.md]: done

File: content/articles/how-to-implement-k-fold-cross-validation/index.md (hunk 2 above)
Comment (Inyrkz) [reply content/articles/how-to-implement-k-fold-cross-validation/index.md]: Done

File: content/articles/how-to-implement-k-fold-cross-validation/index.md (hunk 3 above)
Comment (Inyrkz) [reply content/articles/how-to-implement-k-fold-cross-validation/index.md]: I've made the corrections.

**Comment by Inyrkz** [comment]
Hi @wilkistermumbi. I've made corrections to the article.

File: content/articles/how-to-implement-k-fold-cross-validation/index.md [hunk 4]
```diff
@@ -0,0 +1,417 @@
+### How to Implement K fold Cross-Validation in Scikit-Learn
+
+The most typical strategy in machine learning is to divide a data set into training and validation sets. 70:30 or 80:20 could be the split ratio. It is the holdout method. The problem with this strategy is that we don't know if a high validation accuracy indicates a good model. What if the part of the data we utilized for validation turned out to be a success? Would our model still be accurate if we used a different section of the data set as a validation set? These are some questions that K fold CV answers. The GitHub repo for this project is [here](https://github.com/Inyrkz/breast_cancer/blob/main/k_fold_cv_article_guide.ipynb).
```
Comment (wilkistermumbi) [inline content/articles/how-to-implement-k-fold-cross-validation/index.md:3]: It is good practice to provide the link to your repo before the conclusion section. Not at the top. You first explain the code the provide the link.

File: content/articles/how-to-implement-k-fold-cross-validation/index.md [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+
+### Prerequisites
+- Wisconsin Breast Cancer data set [here](https://www.kaggle.com/uciml/breast-cancer-wisconsin-data)
+- Google Colaboratory or Jupyter Notebook
+- Basic Knowledge of Decision Tree Algorithm [here](https://www.section.io/engineering-education/entropy-information-gain-machine-learning/)
```
Comment (wilkistermumbi) [inline content/articles/how-to-implement-k-fold-cross-validation/index.md:8]: Always begin with, "To follow along with this tutorial, you need to be familiar have:" or a similar sentence. Don't just list them. The tutorial needs to flow smoothly. Also, the sentence, "Basic knowledge of Decision Tree Algorithm here" doesn't make sense. Please correct it accordingly.

File: content/articles/how-to-implement-k-fold-cross-validation/index.md [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+### Introduction
+K fold cross-validation is a superior technique to validate the performance of our model. It evaluates the model using different chunks of the data set as the validation set. 
+
+We divide our data set into K folds. K represents the number of folds into which you want to split your data. If we use 5-folds, the data set divides into five sections. In different iterations, one part becomes the validation set.
+
+![5 Fold Cross-Validation](/engineering-education/how-to-implement-k-fold-cross-validation/5-fold-cv.jpeg)
+
+*[Image Source: Great Learning Blog](https://www.mygreatlearning.com/blog/cross-validation/)*
+
+In the first iteration, we use the first part of the data for validation. We use the other parts of the data set for training as illustrated in the image above.
```
Comment (wilkistermumbi) [inline content/articles/how-to-implement-k-fold-cross-validation/index.md:27]: Consider replacing the second part of the sentence with, "As illustrated in the image above, we use the other parts of the data set for training.	 "

File: content/articles/how-to-implement-k-fold-cross-validation/index.md [hunk 7]
```diff
@@ -0,0 +1,417 @@
...
+            "F1",
+            "F1 Scores in 5 Folds",
+            decision_tree_result["Training F1 scores"],
+            decision_tree_result["Validation F1 scores"])
+```
+
+![F1 Score Plot of Decision Tree Model](/engineering-education/how-to-implement-k-fold-cross-validation/f1-score-1.png)
+
+
+From the visualizations, we notice that the training accuracy, precision, recall, and f1 scores in each fold are 100%. But the validation accuracy, precision, recall and f1 scores are not as high. We call this over-fitting. The model performs admirably on the training data. But not so much on the validation set.
```
Comment (wilkistermumbi) [inline content/articles/how-to-implement-k-fold-cross-validation/index.md:329]: Please replace the first sentence with, "The visualizations show that the training accuracy, precision, recall, and f1 scores in each fold are 100%."

File: content/articles/how-to-implement-k-fold-cross-validation/index.md [hunk 8]
```diff
@@ -0,0 +1,417 @@
...
+            decision_tree_result_2["Validation F1 scores"])
+```
+
+![F1-Score Plot of the Second Decision Tree Model](/engineering-education/how-to-implement-k-fold-cross-validation/f1-score-2.png)
+
+We can see that the validation results of the second model in each fold are better. It has a mean validation accuracy of 93.85% and a mean validation f1 score of 91.69%.
+
+
+### Conclusion
+When training a model on a small data set, the K fold cross-validation technique comes in handy. You may not need to use K fold cross-validation if your data collection is huge. The reason is you have enough records in your validation set to check the machine learning model.  It takes a lot of time to use K fold cross-validation on a large data collection.
```
Comment (wilkistermumbi) [inline content/articles/how-to-implement-k-fold-cross-validation/index.md:411]: Make sure `k-fold` has a hyphen in between. Go through the whole tutorial and edit accordingly

Comment (Inyrkz) [reply content/articles/how-to-implement-k-fold-cross-validation/index.md]: done

File: content/articles/how-to-implement-k-fold-cross-validation/index.md (hunk 7 above)
Comment (Inyrkz) [reply content/articles/how-to-implement-k-fold-cross-validation/index.md]: Done

File: content/articles/how-to-implement-k-fold-cross-validation/index.md (hunk 6 above)
Comment (Inyrkz) [reply content/articles/how-to-implement-k-fold-cross-validation/index.md]: Done

File: content/articles/how-to-implement-k-fold-cross-validation/index.md (hunk 5 above)
Comment (Inyrkz) [reply content/articles/how-to-implement-k-fold-cross-validation/index.md]: Done

File: content/articles/how-to-implement-k-fold-cross-validation/index.md (hunk 4 above)
Comment (Inyrkz) [reply content/articles/how-to-implement-k-fold-cross-validation/index.md]: Done

**Comment by Inyrkz** [comment]
I've made the corrections.

File: content/articles/how-to-implement-k-fold-cross-validation/index.md [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+plot_result(model_name,
+            "F1",
+            "F1 Scores in 5 Folds",
+            decision_tree_result_2["Training F1 scores"],
+            decision_tree_result_2["Validation F1 scores"])
+```
+
+![F1-Score Plot of the Second Decision Tree Model](/engineering-education/how-to-implement-k-fold-cross-validation/f1-score-2.png)
+
+We can see that the validation results of the second model in each fold are better. It has a mean validation accuracy of 93.85% and a mean validation f1 score of 91.69%. The GitHub repo for this project is [here](https://github.com/Inyrkz/breast_cancer/blob/main/k_fold_cv_article_guide.ipynb).
```
Comment (wilkistermumbi) [inline content/articles/how-to-implement-k-fold-cross-validation/index.md:406]: Please replace the last sentence with, "Please find the GitHub repo for this project [here]"

File: content/articles/how-to-implement-k-fold-cross-validation/index.md [hunk 10]
```diff
@@ -0,0 +1,416 @@
+### How to Implement K-fold Cross-Validation in Scikit-Learn
+
+The most typical strategy in machine learning is to divide a data set into training and validation sets. 70:30 or 80:20 could be the split ratio. It is the holdout method. The problem with this strategy is that we don't know if a high validation accuracy indicates a good model. What if the part of the data we utilized for validation turned out to be a success? Would our model still be accurate if we used a different section of the data set as a validation set? These are some questions that K-fold CV answers.
+
+To follow along wwith this tutorial, you need to have:
```
Comment (wilkistermumbi) [inline content/articles/how-to-implement-k-fold-cross-validation/index.md:5]: Always perform a grammar check on your tutorial after adding new information. Also, this section needs to be under the `Prerequisites` section and the sentence "The Wisconsin Breast Cancer data set [here]" is not correct. Please change it to, "The Wisconsin Breast Cancer data set. You can find it [here]"

File: content/articles/how-to-implement-k-fold-cross-validation/index.md [hunk 11]
```diff
@@ -0,0 +1,416 @@
...
+
+We divide our data set into K-folds. K represents the number of folds into which you want to split your data. If we use 5-folds, the data set divides into five sections. In different iterations, one part becomes the validation set.
+
+![5-Fold Cross-Validation](/engineering-education/how-to-implement-k-fold-cross-validation/5-fold-cv.jpeg)
+
+*[Image Source: Great Learning Blog](https://www.mygreatlearning.com/blog/cross-validation/)*
+
+In the first iteration, we use the first part of the data for validation. As illustrated in the image above, we use the other parts of the data set for training.
+
+### Data Preprocessing
```
Comment (wilkistermumbi) [inline content/articles/how-to-implement-k-fold-cross-validation/index.md:28]: Please note that only the first word gets capitalized in a heading.

**Comment by wilkistermumbi** [comment]
@Bayler Please help me add the front matter, PRC line and necessary labels.

**Comment by wilkistermumbi** [comment]
@Inyrkz Please work on the added revisions.

File: content/articles/how-to-implement-k-fold-cross-validation/index.md (hunk 9 above)
Comment (Inyrkz) [reply content/articles/how-to-implement-k-fold-cross-validation/index.md]: done

File: content/articles/how-to-implement-k-fold-cross-validation/index.md (hunk 10 above)
Comment (Inyrkz) [reply content/articles/how-to-implement-k-fold-cross-validation/index.md]: done

File: content/articles/how-to-implement-k-fold-cross-validation/index.md (hunk 11 above)
Comment (Inyrkz) [reply content/articles/how-to-implement-k-fold-cross-validation/index.md]: done

**Comment by wilkistermumbi** [comment]
@Inyrkz  Make sure your tutorial is polished before submitting. Take note of all the revisions we've made in this tutorial so that you can better your next tutorial. @Bayler Please help me add the final review label.

**Comment by Inyrkz** [comment]
Noted. Thank you for your assistance.

**Comment by ahmadmardeni1** [comment]
Great work team. @Inyrkz @Inyrkz

**Comment by hectorkambow** [comment]
@Inyrkz 

Can you please DM via Slack asap 👍  

Thank you

**Comment by Inyrkz** [comment]
Alright

**Pull request merged**